### ISSUE TITLE
New rule: ban unsafe type assertions

### ISSUE THREAD

**Issue by danielnixon** [danielnixon/eslint-plugin-total-functions#19]
These shouldn't be flagged:
```
const foo = "foo" as unknown; // safe
```

```
const foo = "foo" as const; // safe
```

```
const arr = ["foo"];
const foo = arr[0] as string | undefined; // safe
```

But things like this should be:
```
type Foo = {
  readonly bar: number;
  readonly bas: string;
};

const foo = {} as Foo;

foo.bar.toString(); // Explodes at runtime
```

**Comment by danielnixon** [comment]
Background: https://basarat.gitbook.io/typescript/type-system/type-assertion

**Comment by danielnixon** [comment]
The `objectLiteralTypeAssertions` option from the `consistent-type-assertions` rule looks like it handles object literals:

https://github.com/typescript-eslint/typescript-eslint/blob/b5b3be05375c40364d35b887216b56786e33f057/packages/eslint-plugin/docs/rules/consistent-type-assertions.md#objectliteraltypeassertions

But it looks like it wouldn't catch this:

```
type Foo = {
  readonly bar: number;
  readonly bas: string;
};

const foo = {};

const bar = foo as Foo;

bar.bar.toString(); // Explodes at runtime
```

**Comment by danielnixon** [comment]
I've added a `no-unsafe-type-assertion` rule. Now we just need to relax it to not flag "safe" assertions.
